src/ledger/LedgerState.cpp

```diff
@@ -0,0 +1,1228 @@
+// a LedgerEntryReference referring to an offer (with the correct assets) is
+// only valid if it was just returned by loadBestOffer then we know that
+// modifications can only occur at the top of the heap.
+//
+// The consequence of this is that a sequence of calls to loadBestOffer can be
+// efficient if you do not make any intervening calls to LedgerState::load or
+// LedgerState::create that constructed a LedgerEntryReference referring to an
+// offer (with the correct assets). If such an intervening call is made, then
+// the LoadBestOfferContext is destroyed so that it has to be reconstructed at
+// the next call to loadBestOffer, thereby maintaining the constraint.
+LedgerState::StateEntry
+LedgerState::LoadBestOfferContext::loadBestOffer()
+{
+    if (mTop)
+    {
+        // This condition is needed to ensure that the following examples
+        // work correctly:
+        //      Asset a1, a2 = ...;
+        //      LedgerState ls(...);
+        //      ls.loadBestOffer(a1, a2)->erase();
+        //      auto ler = ls.loadBestOffer(a1, a2);
+        //      ...
+        // and:
+        //      Asset a1, a2 = ...;
+        //      LedgerState ls(...);
+        //      *ls.loadBestOffer(a1, a2)->entry() = ...;
+        //      auto ler = ls.loadBestOffer(a1, a2);
+        //      ...
+        // In the first case, we end up resetting mTop so that the next data is
+        // returned. In the second case, mTop still refers to a valid in-memory
+        // value so we push it back onto the heap before loading the next data.
+        // In the second case, it is possible that mTop is returned again (if it
+        // is still the best offer) or that a different offer is returned.
+        if (mTop->valid())
+        {
+            throw std::runtime_error("Previous best offer is still valid");
+        }
+        if (mTop->ignoreInvalid().entry())
+        {
+            auto const& offer = mTop->ignoreInvalid().entry()->data.offer();
+            if (offer.buying == mBuying && offer.selling == mSelling)
+            {
+                mInMemory.push(mTop);
+            }
+        }
+        mTop.reset();
+    }
+
+    while (!mTop)
+    {
+        loadFromDatabaseIfNecessary();
+        if (mInMemory.empty())
+        {
+            if (mFromDatabase.empty())
+            {
+                return nullptr;
+            }
+            else
+            {
+                // In this case, there are offers (with the correct assets) in
+                // memory. It is possible, then, that the best offer in the
+                // database is already loaded in memory. If that is the case, we
+                // must skip the result from the database. Then we repeat the
+                // entire process (not just this branch) to find the next data.
+                mTop = mFromDatabase.front();
+                mFromDatabase.erase(mFromDatabase.begin());
+                if (mLedgerState.isInMemory(LedgerEntryKey(*mTop->entry())))
```

Comment:
I am not sure I understand what is going on here:
isn't `mInMemory` initialized with all ledger states (recursively)?
If so, why would we need to call `isInMemory` like this later on?